pkg/kv/kvserver/logstore/sideload_disk.go

```diff
@@ -285,3 +302,48 @@ func (ss *DiskSideloadStorage) String() string {
 	fmt.Fprintf(&buf, "(%d files)\n", count)
 	return buf.String()
 }
+
+// mkdirAllAndSyncParents creates the given directory and all its missing
+// parents if any. For every newly created directly, it syncs the corresponding
+// parent directory. The directories are created using the provided permissions
+// mask, with the same semantics as in os.MkdirAll.
+//
+// For example, if path is "/x/y/z", and "/x" previously existed, then this func
+// creates "/x/y" and "/x/y/z", and syncs directories "/x" and "/x/y".
+func mkdirAllAndSyncParents(fs vfs.FS, path string, perm os.FileMode) error {
+	// Find the lowest existing directory in the hierarchy.
+	var exists string
+	for dir, parent := path, ""; ; dir = parent {
+		if _, err := fs.Stat(dir); err == nil {
+			exists = dir
+			break
+		} else if !oserror.IsNotExist(err) {
+			return errors.Wrapf(err, "could not get dir info: %s", dir)
+		}
+		parent = fs.PathDir(dir)
+		// NB: not checking against the separator, to be platform-agnostic.
+		if dir == "." || parent == dir { // reached the topmost dir or the root
```

Comment:
~~Instead of special-casing `.` (and what about `..`?), we could simply call `filepath.Abs()` on the path first, which will resolve relative paths and normalize out any `.` and `..` components.~~

Nevermind, I see your comment above that we have to rely on Pebble helpers.

Reply:
I actually have no idea how these helpers behave with paths containing `..`, or in general with non-canonical paths. These are extremely brittle, and I think we should not use them. It would be cleaner to have a type-safe "canonical path" notion which is known to be relative to the data directory root which exists.

Reply:
I think the assumptions are:

- Pebble root directory exists.
- All paths within this directory are well-formed / canonical suffixes of the pebble root path.
- We never walk above the pebble root, so we're effectively always iterating nice canonical substrings.

All these things could be made explicit and constrained within a type-safe helper, so that the space of things we should worry about is reduced. Posted some ideas internally: https://cockroachlabs.slack.com/archives/CAC6K3SLU/p1701873653835429?thread_ts=1700064004.728839&cid=CAC6K3SLU.